File: challenges/2026-09-08-mixer-friend-circles/solution.py
from typing import List


def mixer_friend_circles(n: int, introductions: List[List[int]]) -> List[int]:
    """Group attendees into friend circles and return
    [largestCircleSize, soloCount]."""
    # TODO: implement
    pass


def _run_self_checks() -> None:
    cases = [
        ("example 1", 5, [[0, 1], [1, 2], [3, 4]], [3, 0]),
        ("example 2", 6, [[0, 1], [2, 3]], [2, 2]),
        ("example 3", 1, [], [1, 1]),
        ("edge: duplicate introductions", 3, [[0, 1], [1, 0], [0, 1]], [2, 1]),
        ("edge: full chain merges everyone", 4, [[0, 1], [1, 2], [2, 3]], [4, 0]),
    ]

    for name, n, introductions, expected in cases:
        actual = mixer_friend_circles(n, introductions)
        status = "PASS" if actual == expected else "FAIL"
        print(f"[{status}] {name}: expected={expected!r} actual={actual!r}")


if __name__ == "__main__":
    _run_self_checks()

File: challenges/2026-09-08-mixer-friend-circles/solution.ts
function mixerFriendCircles(n: number, introductions: number[][]): number[] {
  // TODO: implement
  return [];
}

function arraysEqual(a: number[], b: number[]): boolean {
  return a.length === b.length && a.every((v, i) => v === b[i]);
}

function runSelfChecks(): void {
  const cases: [string, number, number[][], number[]][] = [
    ["example 1", 5, [[0, 1], [1, 2], [3, 4]], [3, 0]],
    ["example 2", 6, [[0, 1], [2, 3]], [2, 2]],
    ["example 3", 1, [], [1, 1]],
    ["edge: duplicate introductions", 3, [[0, 1], [1, 0], [0, 1]], [2, 1]],
    ["edge: full chain merges everyone", 4, [[0, 1], [1, 2], [2, 3]], [4, 0]],
  ];

  for (const [name, n, introductions, expected] of cases) {
    const actual = mixerFriendCircles(n, introductions);
    const status = arraysEqual(actual, expected) ? "PASS" : "FAIL";
    console.log(`[${status}] ${name}: expected=${JSON.stringify(expected)} actual=${JSON.stringify(actual)}`);
  }
}

runSelfChecks();
